modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/MergeSortDistributedCacheQueryReducer.java

```diff
@@ -0,0 +1,290 @@
+     * #streams} is sorted to put stream with lowest value on this offset.
+     */
+    private int streamOff;
+
+    /** Compares head of streams to get lowest value at the moment. */
+    private final Comparator<NodePageStream> streamCmp;
+
+    /**
+     * @param rowCmp Comparator to sort query results from different nodes.
+     */
+    public MergeSortDistributedCacheQueryReducer(GridCacheQueryFutureAdapter fut, long reqId,
+        CacheQueryPageRequester fetcher, Collection<ClusterNode> nodes, Comparator<R> rowCmp
+    ) {
+        super(fut, reqId, fetcher, nodes);
+
+        streamsMap = new ConcurrentHashMap<>(nodes.size());
+        streams = (NodePageStream[])Array.newInstance(NodePageStream.class, nodes.size());
+
+        int i = 0;
+
+        for (ClusterNode node : nodes) {
+            streams[i] = new NodePageStream(node.id());
+            streamsMap.put(node.id(), streams[i++]);
+        }
+
+        streamCmp = (o1, o2) -> {
+            if (o1 == o2)
+                return 0;
+
+            if (o1 == null)
+                return -1;
+
+            if (o2 == null)
+                return 1;
+
+            return rowCmp.compare(o1.head, o2.head);
+        };
+    }
+
+    /** {@inheritDoc} */
+    @Override public R next() throws IgniteCheckedException {
+        for (int i = streamOff; i < streams.length; i++) {
+            NodePageStream s = streams[i];
+
+            // Skip if stream has head: on previous next() head extracts from stream[streamOff], others streams do
+            // have head or it is the first run.
+            if (s.head() != null)
+                break;
+
+            if (!s.hasNext()) {
+                // Nullify obsolete stream, move left bound.
+                streamsMap.remove(s.nodeId);
+                streams[i] = null;
+                streamOff++;
+            }
+            else {
+                // Prefetch head value.
+                s.next();
+            }
+        }
+
+        if (finished())
+            throw new NoSuchElementException("No next element. Please, be sure to invoke hasNext() before next().");
+
+        if (first) {
+            first = false;
+
+            Arrays.sort(streams, streamCmp);
+        } else
+            bubbleUp(streams, streamOff, streamCmp);
+
+        return streams[streamOff].get();
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean hasNext() throws IgniteCheckedException {
+        for (int i = streamOff; i < streams.length; i++) {
+            if (streams[i].hasNext())
+                return true;
+
+            // Nullify obsolete stream, move left bound.
+            streamsMap.remove(streams[i].nodeId);
+            streams[i] = null;
+            streamOff++;
+        }
+
+        return false;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void loadPage() {
+        assert !Thread.holdsLock(queueLock());
+
+        List<UUID> nodes;
+
+        synchronized (queueLock()) {
```

Comment:
You use the same lock to synchronize access to streams[] and all queues in all NodeStrems objects.
What is the purpose to use a shared lock?

Reply:
In master branch (single stream) one of purposes of the lock is to guard consistency of `queue` and collections `rcvd`, `subgrid`. As far as I undertstand it's required to avoid excess requests to nodes that already finish query. It could lead to data duplication in final result, as query runs from beginning . Actually as I can see it should be an extremely rare case, but it's possible. Also I don't find any other valuable reasons for this lock.

In implementation with multiple streams we still have shared between streams `rdc` and `subgrid` variables, that has to be in consisten state with all stream queues. So this is a reason to use shared lock there.

Reply:
Actually, I see a way to avoid shared lock for multi-stream reducer. We don't need such collection as `rcvd` in multi-stream, as there a stream per node. So, this code can be simplified. I'll try to implement it.

Reply:
I simplified a code, now NodePageStream doesn't have own logic for control nodes or page loading, it inherited from PageStream. So, reducers doesn't use shared lock now.

But all streams shared the same lock, as it guards consistency between CacheQueryFuture state, queues and rcvd. In one moment only single stream requesting a pages (others do have `head` to use), so it doesn't affect performance, and work the same way as UnorderedReducer does.